Sources/NIO/Selector.swift

```diff
@@ -41,6 +41,192 @@ private extension Optional {
+    static let readHangup = EpollFilterSet(rawValue: 1 << 1)
+    static let input = EpollFilterSet(rawValue: 1 << 2)
+    static let output = EpollFilterSet(rawValue: 1 << 3)
+    static let error = EpollFilterSet(rawValue: 1 << 4)
+
+    init(rawValue: RawValue) {
+        self.rawValue = rawValue
+    }
+}
+
+extension KQueueEventFilterSet {
+    /// Convert NIO's `SelectorEventSet` set to a `KQueueEventFilterSet`
+    init(selectorEventSet: SelectorEventSet) {
+        var thing: KQueueEventFilterSet = .init(rawValue: 0)
+        if selectorEventSet.contains(.read) {
+            thing.formUnion(.read)
+        }
+        if selectorEventSet.contains(.write) {
+            thing.formUnion(.write)
+        }
+
+        // kqueue gives us a `EV_EOF` for the input automatically with `EVFILT_READ` so there's no need to register
+        // for both `EVFILT_EXCEPT` and `EVFILT_READ`.
+
+        // FIXME: this needs to also have .reset
+        if (selectorEventSet.contains(.readEOF) || selectorEventSet.contains(.reset)) && !thing.contains(.read) {
+            thing.formUnion(.except)
+        }
+        self = thing
+    }
+
+    #if os(macOS)
+    /// Calculate the kqueue filter changes that are necessary to transition from `previousKQueueFilterSet` to `self`.
+    /// The `body` closure is then called with the changes necessary expressed as a number of `kevent`.
+    ///
+    /// - parameters:
+    ///    - previousKQueueFilterSet: The previous filter set that is currently registered with kqueue.
+    ///    - fileDescriptor: The file descriptor the `kevent`s should be generated to.
+    ///    - body: The closure that will then apply the change set.
+    func calculateKQueueFilterSetChanges(previousKQueueFilterSet: KQueueEventFilterSet,
+                                         fileDescriptor: CInt,
+                                         _ body: (UnsafeMutableBufferPointer<kevent>) throws -> Void) rethrows {
+        // we only use three filters (EVFILT_READ, EVFILT_WRITE and EVFILT_EXCEPT) so the number of changes would be 3.
+        var keventsHopefullyOnStack = (kevent(), kevent(), kevent())
+        try withUnsafeMutableBytes(of: &keventsHopefullyOnStack) { rawPtr in
+            assert(MemoryLayout<kevent>.size * 3 == rawPtr.count)
+            let keventBuffer = rawPtr.baseAddress!.bindMemory(to: kevent.self, capacity: 3)
+
+            let differences = previousKQueueFilterSet.symmetricDifference(self) // contains all the events that need a change (either need to be added or removed)
+
+            func calculateKQueueChange(event: KQueueEventFilterSet) -> UInt16? {
+                guard differences.contains(event) else {
+                    return nil
+                }
+                return UInt16(self.contains(event) ? EV_ADD : EV_DELETE)
+            }
+
+            var index: Int = 0
+            for (event, filter) in [(KQueueEventFilterSet.read, EVFILT_READ), (.write, EVFILT_WRITE), (.except, EVFILT_EXCEPT)] {
```

Comment:
note to @Lukasa & @normanmaurer . Just in case you were as worried as I was if Swift will optimise this array to be on the stack: It does. I wrote a test program:

```
import Darwin

enum Bla: Int {
    case read = 0xcafe01
    case write = 0xcafe02
    case except = 0xcafe03
}

@inline(never)
func foo() -> (kevent, kevent, kevent) {
    var es: (kevent, kevent, kevent) = (kevent(), kevent(), kevent())
    for (i, e) in [(Bla.read, EVFILT_READ), (.write, EVFILT_WRITE), (.except, EVFILT_EXCEPT)] {
        var ex: kevent = kevent()
        ex.filter = Int16(e)
        ex.flags = i == .read ? 0 : 1
        ex.fflags = 0xdeadbe
        ex.data = i.rawValue
        if i == .read {
            es.0 = ex
        }
        if i == .write {
            es.1 = ex
        }
        if i == .except {
            es.2 = ex
        }
    }
    return es
}

print(foo())
```

which gets really nice assembly:

![screen shot 2018-04-13 at 11 40 38 am](https://user-images.githubusercontent.com/624238/38730708-85d9c3cc-3f0f-11e8-9195-b61754c1900d.png)

so the loop gets fully unrolled etc, awesome job Swift compiler!

Reply:
same seems to happen in the real code. Couldn't spot any calls (except to `initStackObject`) or any loops